syntax = "proto3";

service ModelService {
  rpc train_model (train_request) returns (train_response);
  rpc get_model_types (empty) returns (model_list);
  rpc predict (predict_request) returns (predict_response);
  rpc delete_model (delete_request) returns (delete_response);
  rpc service_status (empty) returns (status_response);
}

message train_request {
  string model_type = 1;
  map<string, string> hyperparams = 2;
}

message train_response {
  string model_id = 1;
}

message predict_request {
  string model_id = 1;
  map<string, float> input_data = 2;
}

message predict_response {
  repeated float prediction = 1;
}

message delete_request {
  string model_id = 1;
}

message delete_response {
  string status = 1;
}

message status_response {
  string status = 1;
  repeated string models = 2;
}

message model_list {
  repeated string model_types = 1;
}

message empty {}